File: server/XSF/Net/NetPacker.cs
//////////////////////////////////////////////////////////////////////////
// 
// 文件：server/XSF/Net/NetPacker.cs
// 作者：Xoen Xie
// 时间：2023/07/01
// 描述：网络协议数据打包器
// 说明：
//
//////////////////////////////////////////////////////////////////////////
#pragma warning disable CS8603

namespace XSF
{
    public sealed class ServerPacker : INetPacker
    {
        private XSFWriter m_Writer;

        public int PackMinLength { get { return 6; } }

        public int PackMaxLength { get { return int.MaxValue; } }

        public ServerPacker()
        {
            m_Writer = new XSFWriter();
            
        }

        public byte[] Read(byte[] recvBuffer, int recvIndex, int nPackageLen, out IMessage message, out ushort nMessageID, out uint nRawID)
        {
            int pbLen = nPackageLen - sizeof(ushort) - sizeof(uint);
            Serilog.Log.Information("Read pbLen=" + pbLen);

            var reader = new XSFReader(recvBuffer, recvIndex + sizeof(uint));
            reader.ReadUShort(out nMessageID);
            reader.ReadUInt(out nRawID);

            Serilog.Log.Information("Read nMessageID=" + nMessageID + ", nRawID=" + nRawID);

            var msg = XSFUtil.GetMessage(nMessageID);
            message = msg.Import(reader.Buffer, reader.CurPos, pbLen);
            
            return null;
        }

        public byte[] Pack(IMessage message)
        {
            m_Writer.Clear();
            byte[] pbData = message.Export();
            Serilog.Log.Information("pack pbData length=" + pbData.Length);
            uint pbLen = (uint)pbData.Length;
            // | 包长(uint 4字节) | 消息ID(ushort 2字节) | rawID(uint 4字节) | pb data |
            uint total = sizeof(ushort) + sizeof(uint) + pbLen;
            Serilog.Log.Information("total = " + total);
            m_Writer.WriteUInt(total);
            m_Writer.WriteUShort(message.ID);
            m_Writer.WriteUInt((uint)message.ID);
            m_Writer.WriteBuffer(pbData);

            byte[] totalData = new byte[m_Writer.Size];
            Serilog.Log.Information("send total = " + totalData.Length);
            Array.Copy(m_Writer.Buffer, totalData, m_Writer.Size);

            return totalData;
        }
    }
}

File: server-cpp/source/xsf/src/net/NetPacker.cpp
//////////////////////////////////////////////////////////////////////////
//
// 文件：source/xsf/src/net/NetPacker.cpp
// 作者：Xoen Xie
// 时间：2020/02/29
// 描述：协议打包器
// 说明：
//
//////////////////////////////////////////////////////////////////////////
#include "NetPakcer.h"
#include "StreamReader.h"

namespace xsf
{
    bool ServerPacker::Read(byte * pRecvBuffer, uint32 nRecvLength, DataResult * pResult)
    {
        StreamReader reader = {"ServerPacker::Read"};
        reader.Init(pRecvBuffer+sizeof(uint32), nRecvLength-sizeof(uint32));
        reader.ReadUint16(pResult->nMessageID);
        reader.ReadUint32(pResult->nRawID);

        pResult->pPBData = reader.Buffer();
        pResult->nPBLength = reader.Size();

        //XSF_INFO("ServerPacker::Read pResult->nMessageID=%u, pResult->nRawID=%u, total length=%u, pb length=%u", pResult->nMessageID, pResult->nRawID, nRecvLength, pResult->nPBLength);

        return true;
    }

    byte * ServerPacker::Pack(IMessage * pMessage, uint32 & nLengthOut)
    {
        m_Writer.Clear();

        uint32 nPbLength = pMessage->GetLength();

        // | 包长(uint 4字节) | 消息ID(ushort 2字节) | rawID(uint 4字节) | pb data |
        uint32 total = sizeof(uint16) + sizeof(uint32) + nPbLength;

        //XSF_INFO("ServerPacker::Pack  id=%u, nPbLength=%u, total=%u", pMessage->GetID(), nPbLength, total + 4);

        m_Writer.WriteUint32(total);
        m_Writer.WriteUint16(pMessage->GetID());
        m_Writer.WriteUint32(pMessage->GetID());
        pMessage->Export(&m_Writer, nPbLength);

        nLengthOut = m_Writer.Size();
        return m_Writer.Buffer();
    }


    byte * ClientPacker::Pack(IMessage * pMessage, uint32 & nLengthOut)
    {
        m_Writer.Clear();

        uint32 nPbLength = pMessage->GetLength();

        // | 包长(uint 4字节) | 消息ID(ushort 2字节) | pb data |
        uint total = sizeof(uint16) + nPbLength;
        m_Writer.WriteUint32(total);
        m_Writer.WriteUint16(pMessage->GetID());
        pMessage->Export(&m_Writer, nPbLength);

        nLengthOut = m_Writer.Size();
        return m_Writer.Buffer();
    }
}